src/iop/highlights.c

```diff
@@ -531,55 +531,110 @@ static void process_lch_bayer(dt_iop_module_t *self, dt_dev_pixelpipe_iop_t *pie
+        // sensor seems prone to this, unlike Bayer, due to its
+        // irregular pattern.
+        int clipped = 1;
+        if (in[0] <= clip)
         {
-          for(int ii = -1; ii <= 1; ii++)
+          for(int offset_j = -2; offset_j <= 0; offset_j++)
           {
-            const float val = in[(size_t)jj * width + ii];
-            const int c = FCxtrans(j+jj, i+ii, roi_in, xtrans);
-            mean[c] += val;
-            cnt[c]++;
-            blend[c] = fmaxf(blend[c], (fminf(post_clip, val) - near_clip) / (post_clip - near_clip));
+            for(int offset_i = -2; offset_i <= 0; offset_i++)
+            {
+              if (clipped) {
+                clipped = 0;
+                for(int jj = offset_j; jj <= offset_j + 2; jj++)
+                {
+                  for(int ii = offset_i; ii <= offset_i + 2; ii++)
+                  {
+                    const float val = in[(ssize_t)jj * roi_in->width + ii];
+                    clipped = (clipped || (val > clip));
+                  }
+                }
+              }
+            }
           }
         }
-        if(blend[0] + blend[1] + blend[2] > 0)
-        { // recover:
-          // options: use max colour and mean blend weight.
-          // const float m = fmaxf(mean[0]/cnt[0], fmaxf(mean[1]/cnt[1], mean[2]/cnt[2]));
-          // const float b = (blend[0] + blend[1] + blend[2])/3.0f;
-          const float m = (mean[0]/cnt[0] + mean[1]/cnt[1] + mean[2]/cnt[2])/3.0f;
-          const float b = fmaxf(fmaxf(blend[0], blend[1]), blend[2]);
-          out[0] = b * m + (1.0f-b) * in[0];
+
+        if(clipped)
+        {
+          float mean[3] = { 0.0f, 0.0f, 0.0f };
+          int cnt[3] = { 0, 0, 0 };
+          float RGBmax[3] = { -FLT_MAX, -FLT_MAX, -FLT_MAX };
+
+          for(int jj = -1; jj <= 1; jj++)
+          {
+            for(int ii = -1; ii <= 1; ii++)
+            {
+              const float val = in[(ssize_t)jj * roi_in->width + ii];
+              const int c = FCxtrans(j+jj, i+ii, roi_in, xtrans);
+              mean[c] += val;
+              cnt[c]++;
+              RGBmax[c] = MAX(RGBmax[c], val);
+            }
+          }
+
+          const float Ro = MIN(mean[0]/cnt[0], clip);
+          const float Go = MIN(mean[1]/cnt[1], clip);
+          const float Bo = MIN(mean[2]/cnt[2], clip);
+
+          const float R = RGBmax[0];
+          const float G = RGBmax[1];
+          const float B = RGBmax[2];
+
+          const float L = (R + G + B) / 3.0f;
+
+          float C = SQRT3 * (R - G);
+          float H = 2.0f * B - G - R;
+
+          const float Co = SQRT3 * (Ro - Go);
+          const float Ho = 2.0f * Bo - Go - Ro;
+
+          if(R != G && G != B)
+          {
+            const float ratio = sqrtf((Co * Co + Ho * Ho) / (C * C + H * H));
+            C *= ratio;
+            H *= ratio;
+          }
+
+          switch(FCxtrans(j, i, roi_out, xtrans))
+          {
+            case 0:
```

Comment:
If we are really going for the speed, i wonder if just computing all three values + outputting just one will be faster, less branching? (like in `process_lch_bayer()`)